File: scripts/r/extraccion_series_macro_y_seleccion_orden_var.R
library(quantmod)
library(tidyquant)
library(tidyverse)
library(gridExtra)
library(ggplot2)
library(dplyr)
library(knitr)
library(vars)

start_date <- "2002-01-01"
end_date <- "2023-01-01"

getSymbols("SERIE_1", src = "FRED", from = start_date, to = end_date)

serie_a <- tq_get("SERIE_A", from = start_date, to = end_date, get = "economic.data")
serie_b <- tq_get("SERIE_B", from = start_date, to = end_date, get = "economic.data")
serie_c <- tq_get("SERIE_C", from = start_date, to = end_date, get = "economic.data")
serie_d <- tq_get("SERIE_D", from = start_date, to = end_date, get = "economic.data")
serie_e <- tq_get("SERIE_E", from = start_date, to = end_date, get = "economic.data")

plot(serie_a$price, type = "l")
lines(serie_b$price)
lines(serie_c$price)
lines(serie_d$price)
lines(serie_e$price)

df <- as.data.frame(cbind(
  fecha = as.Date(serie_a$date),
  var_a = serie_a$price,
  var_b = serie_b$price,
  var_c = serie_c$price,
  var_d = serie_d$price,
  var_e = serie_e$price
))

write.csv(df, "df.csv")

nombres_columnas <- setdiff(names(df), "fecha")

plots <- lapply(nombres_columnas, function(col) {
  ggplot(df, aes(x = fecha, y = .data[[col]])) +
    geom_line() +
    labs(title = col, x = "fecha", y = "variacion")
})

do.call(grid.arrange, c(plots, ncol = 4))

resumen_df <- df %>%
  select(-fecha) %>%
  summarise_all(list(
    media = ~mean(., na.rm = TRUE),
    mediana = ~median(., na.rm = TRUE),
    min = ~min(., na.rm = TRUE),
    max = ~max(., na.rm = TRUE)
  ))

resumen_df$variable <- rownames(resumen_df)
rownames(resumen_df) <- NULL
resumen_df <- resumen_df %>% select(variable, everything())

kable(resumen_df, format = "html")

var_df <- df[, nombres_columnas[1:4]]
exogenas <- df[, nombres_columnas[5:length(nombres_columnas)]]

var_df1 <- var_df %>% mutate(across(everything(), ~ifelse(is.na(.), mean(., na.rm = TRUE), .)))
exo_df <- exogenas %>% mutate(across(everything(), ~ifelse(is.na(.), mean(., na.rm = TRUE), .)))

var_order <- VARselect(var_df1, lag.max = 15, type = "both", exogen = exo_df)

print(var_order$selection)
var_order$criteria

File: scripts/python/extraccion_series_macro_y_seleccion_orden_var.py
import pandas as pd
import matplotlib.pyplot as plt
from pandas_datareader import data as pdr
from statsmodels.tsa.api import VAR

start_date = "2002-01-01"
end_date = "2023-01-01"

serie_1 = pdr.DataReader("SERIE_1", "fred", start_date, end_date)

series_ids = ["SERIE_A", "SERIE_B", "SERIE_C", "SERIE_D", "SERIE_E"]
series = {sid: pdr.DataReader(sid, "fred", start_date, end_date) for sid in series_ids}

plt.figure()
for sid, serie in series.items():
    plt.plot(serie.index, serie.iloc[:, 0], label=sid)
plt.legend()

df = pd.concat(series.values(), axis=1)
df.columns = [f"var_{chr(97 + i)}" for i in range(len(series))]
df = df.reset_index().rename(columns={"DATE": "fecha"})

df.to_csv("df.csv", index=False)

columnas = [c for c in df.columns if c != "fecha"]

fig, axes = plt.subplots(5, 4, figsize=(16, 16))
for ax, col in zip(axes.flatten(), columnas):
    ax.plot(df["fecha"], df[col])
    ax.set_title(col)
fig.tight_layout()

resumen_df = df[columnas].agg(["mean", "median", "min", "max"]).T.reset_index()
resumen_df.columns = ["variable", "media", "mediana", "min", "max"]

var_cols = columnas[:4]
exo_cols = columnas[4:]

var_df = df[var_cols].apply(lambda s: s.fillna(s.mean()))
exo_df = df[exo_cols].apply(lambda s: s.fillna(s.mean()))

modelo = VAR(var_df, exog=exo_df)
seleccion = modelo.select_order(maxlags=15)
print(seleccion.summary())
